catalog/app/utils/AWS/S3.js

```diff
@@ -25,24 +26,65 @@ export const Provider = RT.composeComponent(
   },
 )
 
-export const useS3 = (extra) => {
+export const useS3 = (extra, extend = {}) => {
   const config = Config.use()
   Credentials.use().suspend()
   const props = React.useContext(Ctx)
   // TODO: use cache?
-  return useMemoEq({ ...config, ...props, ...extra }, (cfg) => new S3(cfg))
+  return useMemoEq({ cfg: { ...config, ...props, ...extra }, extend }, (opts) =>
+    Object.assign(new S3(opts.cfg), opts.extend),
+  )
 }
 
 export const use = useS3
 
+const PROXIED = Symbol('proxied')
+
 export const useRequest = (extra) => {
   const cfg = useConfig()
-  const regularClient = useS3(extra)
-  const proxyingClient = useS3({
-    endpoint: cfg.s3Proxy,
-    s3ForcePathStyle: true,
-    ...extra,
-  })
+  const proxyEndpoint = React.useMemo(() => new AWS.Endpoint(cfg.s3Proxy), [cfg.s3Proxy])
+  const customRequestHandler = React.useCallback(
+    (req) => {
+      const b = req.params.Bucket
+      if (b && cfg.shouldProxy(b)) {
+        req.on(
+          'sign',
+          () => {
+            // This is the ultimate hack to get the signing process working.
```

Comment:
This works for now, but would it be simpler to just ask the registry to generate a signed URL (signed by the user's temp credentials) instead of waiting for the retry?